data/transactions/logic/resources.go

```diff
@@ -0,0 +1,328 @@
+// separately, even though they are pointers into the same structure. That
+// prevents dumb attempts to use other fields from the transaction.
+
+func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
+	switch tx.Type {
+	case protocol.PaymentTx:
+		r.fillPayment(&tx.Header, &tx.PaymentTxnFields)
+	case protocol.KeyRegistrationTx:
+		r.fillKeyRegistration(&tx.Header)
+	case protocol.AssetConfigTx:
+		r.fillAssetConfig(&tx.Header, &tx.AssetConfigTxnFields)
+	case protocol.AssetTransferTx:
+		r.fillAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		r.fillAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		r.fillApplicationCall(ep, &tx.Header, &tx.ApplicationCallTxnFields)
+	}
+}
+
+func (r *resources) allows(ep *EvalParams, tx *transactions.Transaction, callerVer, calleeVer uint64) error {
+	switch tx.Type {
+	case protocol.PaymentTx, protocol.KeyRegistrationTx, protocol.AssetConfigTx:
+		// these transactions don't touch cross-product resources, so no error is possible
+		return nil
+	case protocol.AssetTransferTx:
+		return r.allowsAssetTransfer(&tx.Header, &tx.AssetTransferTxnFields)
+	case protocol.AssetFreezeTx:
+		return r.allowsAssetFreeze(&tx.Header, &tx.AssetFreezeTxnFields)
+	case protocol.ApplicationCallTx:
+		return r.allowsApplicationCall(ep, &tx.Header, &tx.ApplicationCallTxnFields, callerVer, calleeVer)
+	default:
+		return fmt.Errorf("unknown inner transaction type %s", tx.Type)
+	}
+}
+
+func (r *resources) fillKeyRegistration(hdr *transactions.Header) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+}
+
+func (r *resources) fillPayment(hdr *transactions.Header, tx *transactions.PaymentTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	r.sharedAccounts[tx.Receiver] = struct{}{}
+	if !tx.CloseRemainderTo.IsZero() {
+		r.sharedAccounts[tx.CloseRemainderTo] = struct{}{}
+	}
+}
+
+func (r *resources) fillAssetConfig(hdr *transactions.Header, tx *transactions.AssetConfigTxnFields) {
+	r.sharedAccounts[hdr.Sender] = struct{}{}
+	if id := tx.ConfigAsset; id != 0 {
+		r.sharedAsas[id] = struct{}{}
+	}
+	// We don't need to read the special addresses, so they don't go in.
+}
+
+func (r *resources) fillAssetTransfer(hdr *transactions.Header, tx *transactions.AssetTransferTxnFields) {
+	id := tx.XferAsset
+	r.sharedAsas[id] = struct{}{}
+	r.shareAccountAndHolding(hdr.Sender, id)
+	r.shareAccountAndHolding(tx.AssetReceiver, id)
+
+	if !tx.AssetSender.IsZero() {
+		r.shareAccountAndHolding(tx.AssetSender, id)
+	}
+
+	if !tx.AssetCloseTo.IsZero() {
+		r.shareAccountAndHolding(tx.AssetCloseTo, id)
+	}
+}
+
+// allowsHolding checks if a holding is available under the txgroup sharing rules
+func (r *resources) allowsHolding(addr basics.Address, ai basics.AssetIndex) bool {
+	if _, ok := r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: ai}]; ok {
+		return true
+	}
+	// All holdings of created assets are available
```

Comment:
I think a similar additional check should be added to allow all holdings by newly created app accounts (same applies to `allowsLocals` too).

Otherwise, a newly created app account would be very limited, e.g. can't opt into assets or be passed to pre-v9 apps as a foreign account

Reply:
Added.  Have a look. I don't like the current arrangement of where the functionality is located (some in eval.go, some in resources.go) because no the interaction between those files is more fluid.  But I think it it's correct now. (Needs more testing, especially of locals, compared to assets)